pkg/nativeimgutil/nativeimgutil.go

```diff
@@ -0,0 +1,156 @@
+			return err
+		}
+		return convertRawToRaw(source, dest, size)
+	case qcow2.Type:
+		if !allowSourceWithBackingFile {
+			q, ok := srcImg.(*qcow2.Qcow2)
+			if !ok {
+				return fmt.Errorf("unexpected qcow2 image %T", srcImg)
+			}
+			if q.BackingFile != "" {
+				return fmt.Errorf("qcow2 image %q has an unexpected backing file: %q", source, q.BackingFile)
+			}
+		}
+	default:
+		logrus.Warnf("image %q has an unexpected format: %q", source, t)
+	}
+	if err = srcImg.Readable(); err != nil {
+		return fmt.Errorf("image %q is not readable: %w", source, err)
+	}
+
+	// Create a tmp file because source and dest can be same.
+	destTmpF, err := os.CreateTemp(filepath.Dir(dest), filepath.Base(dest)+".lima-*.tmp")
+	if err != nil {
+		return err
+	}
+	destTmp := destTmpF.Name()
+	defer os.RemoveAll(destTmp)
+	defer destTmpF.Close()
+
+	// Copy
+	srcImgR := io.NewSectionReader(srcImg, 0, srcImg.Size())
+	const bufSize = 1024 * 1024
+	if copied, err := copySparse(destTmpF, srcImgR, bufSize); err != nil {
+		return fmt.Errorf("failed to call copySparse(), bufSize=%d, copied=%d: %w", bufSize, copied, err)
+	}
+
+	// Resize
+	if size != nil {
+		if err = MakeSparse(destTmpF, *size); err != nil {
+			return err
+		}
+	}
+	if err = destTmpF.Close(); err != nil {
+		return err
+	}
+
+	// Rename destTmp into dest
+	if err = os.RemoveAll(dest); err != nil {
+		return err
+	}
+	return os.Rename(destTmp, dest)
+}
+
+func convertRawToRaw(source, dest string, size *int64) error {
+	if source != dest {
+		// continuity attempts clonefile
+		if err := fs.CopyFile(dest, source); err != nil {
+			return fmt.Errorf("failed to copy %q into %q: %w", source, dest, err)
+		}
+	}
+	if size != nil {
+		destF, err := os.OpenFile(dest, os.O_RDWR, 0644)
+		if err != nil {
+			return err
+		}
+		defer destF.Close()
+		if err = MakeSparse(destF, *size); err != nil {
+			return err
+		}
+		if err = destF.Close(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func copySparse(w *os.File, r io.Reader, bufSize int64) (int64, error) {
+	var n int64
+	zeroBuf := make([]byte, bufSize)
+	buf := make([]byte, bufSize)
+	var eof bool
+	for !eof {
+		rN, rErr := r.Read(buf)
+		if rErr != nil {
+			eof = errors.Is(rErr, io.EOF)
+			if !eof {
+				return n, fmt.Errorf("failed to read: %w", rErr)
+			}
+		}
+		// TODO: qcow2reader should have a method to notify whether buf is zero
+		if bytes.Equal(buf, zeroBuf) {
+			if _, sErr := w.Seek(int64(rN), os.SEEK_CUR); sErr != nil {
+				return n, fmt.Errorf("failed seek: %w", sErr)
+			}
+			// no need to ftruncate here
+			n += int64(rN)
+		} else {
+			wN, wErr := w.Write(buf)
+			if wN > 0 {
+				n += int64(wN)
+			}
+			if wErr != nil {
+				return n, fmt.Errorf("failed to read: %w", wErr)
+			}
+			if wN != rN {
+				return n, fmt.Errorf("read %d, but wrote %d bytes", rN, wN)
+			}
+		}
+	}
+	return n, nil
+}
+
+func MakeSparse(f *os.File, n int64) error {
+	if _, err := f.Seek(n, os.SEEK_SET); err != nil {
```

Comment:
os.SEEK_SET is deprecated.

We could use the io.SeekStart

Reply:
Fixed